bigtable/metrics.go

```diff
@@ -0,0 +1,479 @@
+	monitoredResLabelKeyZone     = "zone"
+
+	// Metric labels
+	metricLabelKeyAppProfile         = "app_profile"
+	metricLabelKeyMethod             = "method"
+	metricLabelKeyOperationStatus    = "status"
+	metricLabelKeyStreamingOperation = "streaming"
+	metricLabelKeyClientName         = "client_name"
+	metricLabelKeyClientUID          = "client_uid"
+
+	// Metric names
+	metricNameOperationLatencies = "operation_latencies"
+	metricNameAttemptLatencies   = "attempt_latencies"
+	metricNameRetryCount         = "retry_count"
+	metricNameServerLatencies    = "server_latencies"
+)
+
+var (
+	// duration between two metric exports
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName            = fmt.Sprintf("cloud.google.com/go/bigtable v%v", internal.Version)
+	builtInEnabledDefault = true
+
+	// All the built-in metrics have same attributes except 'status' and 'streaming'
+	// These attributes need to be added to only few of the metrics
+	builtinMetrics = map[string]metricInfo{
+		metricNameOperationLatencies: {
+			desc:       "Total time until final operation success or failure, including retries and backoff.",
+			metricType: otelmetric.InstrumentKindHistogram,
+			unit:       "ns",
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+				metricLabelKeyStreamingOperation,
+			},
+		},
+		metricNameAttemptLatencies: {
+			desc:       "Client observed latency per RPC attempt.",
+			metricType: otelmetric.InstrumentKindHistogram,
+			unit:       "ns",
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+				metricLabelKeyStreamingOperation,
+			},
+		},
+		metricNameServerLatencies: {
+			desc:       "The latency measured from the moment that the RPC entered the Google data center until the RPC was completed.",
+			metricType: otelmetric.InstrumentKindHistogram,
+			unit:       "ns",
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+				metricLabelKeyStreamingOperation,
+			},
+		},
+		metricNameRetryCount: {
+			desc:       "The number of additional RPCs sent after the initial attempt.",
+			metricType: otelmetric.InstrumentKindCounter,
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+			},
+		},
+	}
+
+	noOpRecordFn = func() {}
+)
+
+type metricInfo struct {
+	desc                 string
+	metricType           otelmetric.InstrumentKind
+	unit                 string
+	additionalAttributes []string
+}
+
+type builtInMetricInstruments struct {
+	distributions map[string]metric.Float64Histogram // Key is metric name e.g. operation_latencies
+	counters      map[string]metric.Int64Counter     // Key is metric name e.g. retry_count
+}
+
+// Generates unique client ID in the format go-<random UUID>@<>hostname
+func generateClientUID() (string, error) {
+	hostname := "localhost"
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", err
+	}
+	return "go-" + uuid.NewString() + "@" + hostname, nil
+}
+
+type metricsConfigInternal struct {
```

Comment:
This isnt a Config ... in cbt Configs are value classes that allow end users to customize the client behavior.
This is struct has methods like recordAttemptCompletion, which suggest that its not just a value object.

Maybe it should be called something like CbtMetricsRecorder?

Reply:
Also should there be an interface with 2 impls:
- A real impl that forwards the metrics
- A no-op impl that drops the metrics

Reply:
Updated the names and added implementations